FIX stderr variable increasing to infinite string on livestreaming

christophjohannsdotter opened the pull request:
The var 'stderr' appends all data, that comes on process event on 'data'. That causes a very long string on livestreaming, since 3-5 lines are added per second.

On the on 'progress' event, there are RegEx made on this string and i noticed, that this causes high cpu load on long strings, especially the raspberry pi has a cpuload of 100% after 2-3 hours of livestreaming with the fluent ffmpeg module. A v8 dump told me, that the load is caused by javascript RegEx method calls, so i started to search for the reason. 

I changed the code to only use the last lines of stderr if 50 lines of stderr output have been reached. Since the codecData is sent on these 50 lines of stderr output, i store the first 50 lines in an extra process-variable, to make dynamic adding of the 'codecData' event also possible. 

The result on my raspberry pi is, that the fluent-ffmpeg prcoess does not increase the cpu-load any more.

njoyard commented:
Good idea, however this should be optional. You should be able to get the whole stderr output if needed, and maybe the automatic line limiting should only occur when streaming... Also, the stderr parsing logic for codecData/progress events can probably be optimized.

christophjohannsdotter commented:
Thanks for the reply. I'm not sure if any process really needs the whole stderr since it keeps only a \n divided string with currentFps, currentKbps and stuff, that is not useful anymore after it has been sent, but i might forget any case, in which this could be usefull (perhaps if running statistics on prior values of currentFps). 

Another way to solve the problem might be to cut each progressline after it has been parsed like 

```javascript
//do regex to divide into lines here
var lines = stderr.split(nlRegexp);
//pass lines now
utils.extractProgress(self, lines, duration);
//remove last line since it has been parsed now
lines.splice(-1,1).join("\n"); 

If you still think this should be optional for livestreaming, how would you do that? like
```javascript
command = new FfmpegCommand(someUrl);
command.setLivestream(true)

? 

Another question since i am new to creating pull requests, do i have to create a new pull request to update the code i submitted? And can i link this commit to this thread?

njoyard commented:
stdout and stderr are passed to the 'end' and 'error' handlers, so they are used by fluent-ffmpeg users. Changing what those handlers receive is an incompatible API change, so it's a no-no.

Additionnaly there are plans already to add new events regarding handling of ffmpeg output.

As for the pull request, you may do either one of those:
- add new commits to the existing pull request by simply pushing new commits on your FIX_stderr-variable_of_infinite_growth branch
- close this pull request and open a new one

christophjohannsdotter commented:
Thanks for your feedback, i agree with you that changing stderr or stdout in general is no good idea. I updated the code to make the linelimit optional. I also updated the README to publish the new option.

njoyard commented:
Sorry I let this without news for that long :)

If you have some time could you please :
- maybe write a test or two
- add an option for how many "start" lines we want to keep (for example in some cases we only want to keep the *last* 50 lines)

If you don't have time I'll merge this and add "feature request" issues instead.

overlookmotel commented:
I've run into the same problem when using filters with a lot of verbose output. For example, the `bbox` filter outputs a line to stderr for every frame of video. When analysing a 5-minute video with `bbox`, the process soon grinds to a halt due to storing the entire stderr output in memory.

I've written a work-around that does the following:

1. allows you to completely turn off stderr capturing, but maintains progress reporting etc
2. parses progress and startup stderr lines without using a regexp on the entire stderr string every time.

NB You can still get notified of stderr output by listening on the `stderr` event (see PR #467).

However, I haven't submitted it as a PR because I'm confused about what the default behaviour is meant to be...

There is an option `captureStderr` which defaults to `false`. The presence of this option suggests to me that the default behaviour should be not to capture stderr. But in practice it does, and many people probably rely on getting the stderr output returned at the end of the process.

Here's a suggestion:

* Change default for `captureStderr` option to `true`
* If `captureStderr` option is NOT true, do not capture stderr (user can use `.on('stderr', function() { ... })` to get the stderr output line by line if they need it)

If this sounds like a good idea, I'll submit my patch as a PR. That'd I think solve the same problem this PR is trying to solve.

What do you think @njoyard and @christophjohannsdotter?

njoyard commented:
`captureStderr` isn't a user option. When refactoring the code I needed to add `captureStdout` to ffmpeg._spawnFfmpeg options in order to differentiate stream output from file output handling, and for homogeneity's sake I also added `captureStderr`.

I'm currently reworking on this PR to have a 100 line limit by default. I'm getting rid of the "keep the X first lines" thingy because if you need that you can now do it with the 'stderr' event. But I'd like to keep those active by default to help newbies debug issues.

njoyard commented:
But if you want you can make a PR that adds options that force **not** capturing stdout/stderr :)

christophjohannsdotter commented:
hi guys, 

for our project https://github.com/datarhei/restreamer that heavily uses the great node-fluent-ffmpeg module, we needed to have a solution, that reduces the amount of lines in the stderr to a fixed limit, since the raspberry pi had a 100% cpu load after a while. 
What i did was adding an extra-option to the constructor, you can see it here:

https://github.com/datarhei/restreamer/blob/master/src/classes/Restreamer.js#L288

outputLineLimit:1

We needed to have at least one line in the stderr to have the progressData of the stream, so disabling the stderr output would not be a solution for us.

You can find the fork here:

https://github.com/datarhei/node-fluent-ffmpeg

If you think, this could be a good solution, than i will recreate the pull request.

overlookmotel commented:
@njoyard OK cool. That's how I wrote it originally - adding `noStdErr` and `noStdOut` options to supress output. I'll submit a PR.

What branch should PRs be based on? `master` or `merges`?

overlookmotel commented:
PS In my opinion it's not necessary to cut stderr output to last N lines by default, which would be a breaking change.

I think the common use case could be solved with a `noStdErr` option, which would fix the memory issues for ffmpeg commands which produce a lot of output (relatively uncommon). And if a user needs something more specialist, like a ring buffer of the last N lines, they can always implement it using the `stderr` event.

njoyard commented:
Always base on master, merges is just some WIP helper branch.

After thinking about it, I think the current implementation of stdout/stderr storage is flawed : it allows uncontrolled grow of memory usage in its default state for some use cases (streaming live input or using something that outputs lots of messages). I now see this as a bug. Especially since stdout/stderr storage were implemented mainly for debug purposes (progress/codecData don't need *storage*).

I'm going to implement a ring buffer with a default limit around 100 lines or so. I think the number of actual use cases this breaks will be quite low, as you only get the contents from those buffers in the 'end' / 'error' handlers. You'll still be able to raise the limit if necessary or use the 'stderr' event for anything more complicated.

njoyard commented:
...BTW i wrote "i'm going to" in a very imperative way, but I'm still open to arguments against it :)

overlookmotel commented:
@njoyard Ah OK. Well I tend to agree with you! But this will break backward compatibility which you earlier said was a no-no, so I wasn't going to argue.

Give me a few minutes to publish my PR. I think there may be something there that you like - it caches only lines which could be an error, and discards them as soon as it becomes apparent that they're not. So if an error is thrown, the user can always get the full error output back. But that's without caching *all* stderr output, and there's no set limit to how many lines an error can be.

Just wrangling with SSH keys (evidently haven't committed to GitHub for a while!), and I'll have the PR up.

njoyard commented:
Yeah I know, but I changed my mind... and in France we say that only dumbasses never change their minds :)

I'm stopping for a bit, going to eat then sleep, first !

overlookmotel commented:
You Frenchies are so wise :) Anyway, here it is: PR https://github.com/fluent-ffmpeg/node-fluent-ffmpeg/pull/511

njoyard commented:
Rewrote in 96d9ed543d337845d463b48d56a4dd7db8b66548
